eip/EIPS/eip-aa-rules.md

```diff
@@ -0,0 +1,351 @@
+* `*CALL` opcodes:
+    * **[OP-061]** `CALL` with `value` is forbidden. The only exception is a call to the `EntryPoint` described above.
+    * **[OP-062]** Precompiles:
+        * Only allowed the core 9 precompiles.\
+          Specifically the computation precompiles that do not access anything in the blockchain state or environment.
+
+### Code Rules
+
+* **[COD-010]** Between the first and the second validations, the `EXTCODEHASH` value of any visited address,
+entity or referenced library, may not be changed.\
+If the code is modified, that entity is throttled down. \
+If the code of a library was modified, then the entity that used that library is throttled.
+* **[COD-020]**: Code Modification Throttling: \
+In order to prevent "mass invalidation" by a self-modifying code, if the contract code is modified,
+it is throttled down and gains reputation as long as it does not change.
+    * Maintain a list of `{address, codehash, last modified}`
+    * Every contract accessed by `*CALL` or `EXTCODE*` during validation is saved into this list.
+    * If during any validation we find that a code was changed, the entity using it is marked as "throttled".
+        * Allow only `THROTTLED_ENTITY_MEMPOOL_COUNT` transactions in the mempool.
+    * If the `UserOperation` is included on-chain, drop entry from the list.
+* **[COD-030]** Once `SETCODE` opcode ([EIP-6913](https://eips.ethereum.org/EIPS/eip-6913)) or a similar feature is added:
+    * Any entity code except the account itself  (`paymaster`, `factory`, and any referenced library) may include it only if that entity is staked.
+    * This rule applies to the entire executed code section - including external contract code called by
+      `*CALL` during validation -  not just the code that was actually visited during validation.
+
+### Storage rules.
+
+The permanent storage access with `SLOAD` and `SSTORE` instructions within each phase are limited as follows:
+
+* **[STO-010]** Access to the "account" storage is always allowed.
+* Access to associated storage of the account in an external (non-entity contract) is allowed if:
+    * **[STO-021]**  The account already exists
+    * **[STO-022]**  There is an `initCode` and the `factory` contract is staked.
+* If the entity (`paymaster`, `factory`) is staked, then it is also allowed:
+    * **[STO-031]** Access the entity's own storage.
+    * **[STO-032]** Access a storage slot in any non-entity contract that is associated with the entity.
+* **[STO-040]** `UserOperation` may not use an entity address (`factory`/`paymaster`/`aggregator`) that is used as an "account" in another `UserOperation` in the mempool. \
+This means that a `Paymaster` and `Factory` contracts cannot practically be an "account" contract as well.
+* **[STO-041]** A contract whose Associated Storage slot is accessed during a `UserOperation` may not be an address of a "sender" in another `UserOperation` in the mempool.
```

Comment:
`STO-040` and `STO-041` seem like they are going to be very difficult to enforce in the P2P network since each bundler will have a different view of the mempool.

Its also dependent on the ordering upon which UOs enter the mempool. Unless each time we add a UO to the pool we do a scan over every operation in the pool for storage accesses.

Reply:
These rules come more to protect the bundler itself, and not the entire mempool.
That is, you may ignore those rules (making you easier to be DoS - attacked). 
Also, you're not expected to mark a sender of such userops as a "spammer"
We should add better definition of these type of rules (consensus rules and bundler-specific rules)